File: construction_v1/report/employer_status_analysis/employer_status_analysis.js
/**
 * Employer Status Analysis - Script Report Frontend
 *
 * Production-ready ERPNext Script Report with filters, charts, and WorkCom AI integration.
 * Uses Employer doctype as the data source.
 */

frappe.query_reports["Employer Status Analysis"] = {
    filters: [
        {
            fieldname: "period_type",
            label: __("Period Type"),
            fieldtype: "Select",
            options: construction_v1.report_utils.period_type_options,
            default: "Monthly",
            reqd: 1,
            on_change: function () {
                construction_v1.report_utils.handle_period_change();
            }
        },
        {
            fieldname: "date_from",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.month_start(),
            reqd: 1
        },
        {
            fieldname: "date_to",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 1
        },
        {
            fieldname: "filter_by_date",
            label: __("Filter by Registration Date"),
            fieldtype: "Check",
            default: 0
        },
        {
            fieldname: "branch",
            label: __("Branch"),
            fieldtype: "Link",
            options: "Branch"
        },
        {
            fieldname: "status",
            label: __("Status"),
            fieldtype: "Select",
            options: "\nActive\nInactive\nSuspended\nPending Verification\nBlacklisted"
        }
    ],

    onload: function (report) {
        // Add WorkCom AI button
        report.page.add_inner_button(__("Ask WorkCom"), function () {
            show_WorkCom_dialog(report);
        }, __("AI Insights"));

        // Add additional chart buttons
        report.page.add_inner_button(__("Status Chart"), function () {
            show_status_chart();
        }, __("Charts"));

        report.page.add_inner_button(__("Branch Chart"), function () {
            show_branch_chart();
        }, __("Charts"));

        report.page.add_inner_button(__("Trend Chart"), function () {
            show_trend_chart();
        }, __("Charts"));
    },

    formatter: function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "status" && data) {
            let status = data.status || "";
            if (status === "Active") {
                value = `<span class="indicator-pill green">${value}</span>`;
            } else if (status === "Inactive") {
                value = `<span class="indicator-pill grey">${value}</span>`;
            } else if (status === "Suspended") {
                value = `<span class="indicator-pill orange">${value}</span>`;
            } else if (status === "Pending Verification") {
                value = `<span class="indicator-pill blue">${value}</span>`;
            } else if (status === "Blacklisted") {
                value = `<span class="indicator-pill red">${value}</span>`;
            }
        }

        return value;
    }
};

function show_WorkCom_dialog(report) {
    let d = new frappe.ui.Dialog({
        title: __("Ask WorkCom - AI Analytics Assistant"),
        fields: [
            {
                fieldname: "query",
                label: __("Your Question"),
                fieldtype: "Small Text",
                reqd: 1,
                placeholder: __("e.g., What trends do you see in employer registrations?")
            },
            {
                fieldname: "response_section",
                fieldtype: "Section Break",
                label: __("WorkCom's Response")
            },
            {
                fieldname: "response",
                fieldtype: "HTML",
                options: '<div class="WorkCom-response" style="min-height:100px;padding:10px;background:#f5f7fa;border-radius:4px;"><em>Ask a question to get AI-powered insights...</em></div>'
            }
        ],
        primary_action_label: __("Ask WorkCom"),
        primary_action: function (values) {
            let $response = d.$wrapper.find(".WorkCom-response");
            $response.html('<div class="text-muted"><i class="fa fa-spinner fa-spin"></i> WorkCom is thinking...</div>');

            frappe.call({
                method: "construction_v1.construction_v1.report.employer_status_analysis.employer_status_analysis.get_ai_insights",
                args: {
                    filters: JSON.stringify(report.get_filter_values()),
                    query: values.query
                },
                callback: function (r) {
                    let answer = (r && r.message && r.message.insights) || "No response received.";
                    $response.html(`<div style="white-space:pre-wrap;">${answer}</div>`);
                },
                error: function () {
                    $response.html('<div class="text-danger">Error getting AI insights. Please try again.</div>');
                }
            });
        }
    });
    d.show();
}

function show_status_chart() {
    frappe.call({
        method: "construction_v1.construction_v1.report.employer_status_analysis.employer_status_analysis.get_status_chart",
        callback: function (r) {
            if (r && r.message) {
                show_chart_dialog(__("Employers by Status"), r.message);
            }
        }
    });
}

function show_branch_chart() {
    frappe.call({
        method: "construction_v1.construction_v1.report.employer_status_analysis.employer_status_analysis.get_branch_chart",
        callback: function (r) {
            if (r && r.message) {
                show_chart_dialog(__("Top 10 Branches by Employer Count"), r.message);
            }
        }
    });
}

function show_trend_chart() {
    frappe.call({
        method: "construction_v1.construction_v1.report.employer_status_analysis.employer_status_analysis.get_trend_chart",
        args: { months: 6 },
        callback: function (r) {
            if (r && r.message) {
                show_chart_dialog(__("Employer Registration Trend (Last 6 Months)"), r.message);
            }
        }
    });
}

function show_chart_dialog(title, chart_data) {
    let d = new frappe.ui.Dialog({
        title: title,
        size: "large"
    });

    d.show();

    // Render chart in dialog body
    let chart_container = $('<div class="chart-container" style="height:400px;"></div>');
    d.$body.append(chart_container);

    new frappe.Chart(chart_container[0], {
        data: chart_data.data,
        type: chart_data.type || "bar",
        height: 350,
        colors: chart_data.colors || ["#5e64ff"]
    });
}

File: construction_v1/report/employer_status_analysis/employer_status_analysis.py
"""
Employer Status Analysis - Script Report

Production-ready Script Report for analyzing employer status distributions,
registration trends, and branch associations using the Employer doctype.

Includes WorkCom AI integration for intelligent insights.
"""

import json
from typing import Any, Dict, List, Optional, Tuple

import frappe
from frappe.utils import getdate, add_months, get_first_day, get_last_day
from construction_v1.report.report_utils import get_period_dates


def execute(filters: Optional[Dict[str, Any]] = None) -> Tuple:
    """Main entry point for Script Report."""
    filters = frappe._dict(filters or {})

    # Ensure date filters
    get_period_dates(filters)

    columns = get_columns()
    data, counts = get_data(filters)
    chart = get_chart_data(counts)
    report_summary = get_report_summary(counts)

    return columns, data, None, chart, report_summary, False


def get_columns() -> List[Dict[str, Any]]:
    """Define report columns based on Employer doctype fields."""
    return [
        {"fieldname": "name", "label": "Employer ID", "fieldtype": "Link", "options": "Employer", "width": 140},
        {"fieldname": "employer_name", "label": "Employer Name", "fieldtype": "Data", "width": 200},
        {"fieldname": "employer_code", "label": "Employer Code", "fieldtype": "Data", "width": 120},
        {"fieldname": "status", "label": "Status", "fieldtype": "Data", "width": 130},
        {"fieldname": "branch_name", "label": "Branch", "fieldtype": "Link", "options": "Branch", "width": 150},
        {"fieldname": "branch_code", "label": "Branch Code", "fieldtype": "Data", "width": 100},
        {"fieldname": "creation", "label": "Registration Date", "fieldtype": "Date", "width": 120},
    ]


def get_data(filters: frappe._dict) -> Tuple[List[Dict[str, Any]], Dict[str, int]]:
    """Fetch and aggregate employer data from Employer doctype."""
    # Build SQL conditions
    conditions = ["1=1"]
    values = {}

    # Status filter
    if filters.get("status"):
        conditions.append("status = %(status)s")
        values["status"] = filters.status

    # Branch filter
    if filters.get("branch"):
        conditions.append("branch_name = %(branch)s")
        values["branch"] = filters.branch

    # Date range filter on creation
    if filters.get("filter_by_date"):
        df = getdate(filters.date_from)
        dt = getdate(filters.date_to)
        conditions.append("DATE(creation) BETWEEN %(df)s AND %(dt)s")
        values["df"] = df
        values["dt"] = dt

    where_clause = " AND ".join(conditions)

    # Fetch employers from Employer table
    rows = frappe.db.sql(f"""
        SELECT
            name, employer_name, employer_code, status, branch_name, branch_code,
            DATE(creation) as creation
        FROM `tabEmployer`
        WHERE {where_clause}
        ORDER BY creation DESC
        LIMIT 5000
    """, values, as_dict=True)

    # Build data rows and count statuses
    data = []
    counts = {
        "total": 0,
        "active": 0,
        "inactive": 0,
        "suspended": 0,
        "pending": 0,
        "blacklisted": 0,
    }

    for r in rows:
        counts["total"] += 1
        status = (r.get("status") or "").strip()

        # Count by status
        if status == "Active":
            counts["active"] += 1
        elif status == "Inactive":
            counts["inactive"] += 1
        elif status == "Suspended":
            counts["suspended"] += 1
        elif status == "Pending Verification":
            counts["pending"] += 1
        elif status == "Blacklisted":
            counts["blacklisted"] += 1

        data.append({
            "name": r.get("name"),
            "employer_name": r.get("employer_name"),
            "employer_code": r.get("employer_code"),
            "status": status or "Unknown",
            "branch_name": r.get("branch_name"),
            "branch_code": r.get("branch_code"),
            "creation": r.get("creation"),
        })

    return data, counts


def get_chart_data(counts: Dict[str, int]) -> Dict[str, Any]:
    """Build chart data for status distribution."""
    return {
        "data": {
            "labels": ["Active", "Inactive", "Suspended", "Pending", "Blacklisted"],
            "datasets": [{
                "name": "Employers",
                "values": [
                    counts.get("active", 0),
                    counts.get("inactive", 0),
                    counts.get("suspended", 0),
                    counts.get("pending", 0),
                    counts.get("blacklisted", 0),
                ]
            }]
        },
        "type": "pie",
        "colors": ["#28a745", "#6c757d", "#ffc107", "#17a2b8", "#dc3545"],
    }


def get_report_summary(counts: Dict[str, int]) -> List[Dict[str, Any]]:
    """Build report summary cards."""
    return [
        {"value": counts.get("total", 0), "label": "Total Employers", "datatype": "Int"},
        {"value": counts.get("active", 0), "label": "Active", "datatype": "Int", "indicator": "green"},
        {"value": counts.get("inactive", 0), "label": "Inactive", "datatype": "Int", "indicator": "grey"},
        {"value": counts.get("suspended", 0), "label": "Suspended", "datatype": "Int", "indicator": "orange"},
        {"value": counts.get("pending", 0), "label": "Pending", "datatype": "Int", "indicator": "blue"},
        {"value": counts.get("blacklisted", 0), "label": "Blacklisted", "datatype": "Int", "indicator": "red"},
    ]


def get_distribution_maps() -> Tuple[Dict[str, int], Dict[str, int]]:
    """Get status and branch distributions via SQL aggregation."""
    status_map = {}
    for row in frappe.db.sql(
        "SELECT IFNULL(status, 'Unknown') AS s, COUNT(*) AS cnt FROM `tabEmployer` GROUP BY s",
        as_dict=True
    ):
        status_map[row.s] = int(row.cnt)

    branch_map = {}
    for row in frappe.db.sql(
        "SELECT IFNULL(branch_name, 'Unassigned') AS b, COUNT(*) AS cnt FROM `tabEmployer` GROUP BY b ORDER BY cnt DESC LIMIT 10",
        as_dict=True
    ):
        branch_map[row.b] = int(row.cnt)

    return status_map, branch_map


@frappe.whitelist()
def get_ai_insights(filters: str, query: str) -> Dict[str, Any]:
    """Return WorkCom-style insights for the Employer Status Analysis report."""
    filters = frappe._dict(json.loads(filters) if isinstance(filters, str) else filters or {})

    # Ensure dates
    if not filters.get("date_from") or not filters.get("date_to"):
        filters.date_to = getdate()
        filters.date_from = frappe.utils.add_days(filters.date_to, -29)

    # Get current data
    _, counts = get_data(filters)
    status_map, branch_map = get_distribution_maps()

    context = {
        "window": {
            "period_type": filters.get("period_type", "Custom"),
            "from": str(filters.date_from),
            "to": str(filters.date_to),
        },
        "current": {
            "total": counts.get("total", 0),
            "active": counts.get("active", 0),
            "inactive": counts.get("inactive", 0),
            "suspended": counts.get("suspended", 0),
            "pending": counts.get("pending", 0),
            "blacklisted": counts.get("blacklisted", 0),
        },
        "distributions": {
            "status": status_map,
            "branch": branch_map,
        },
    }

    try:
        from construction_v1.services.enhanced_ai_service import EnhancedAIService
        ai = EnhancedAIService()
        answer = ai.generate_employer_status_report_insights(query=query, context=context)
        return {"insights": answer}
    except Exception:
        frappe.log_error(frappe.get_traceback(), "Employer Status Analysis AI Insights Error")
        return {
            "insights": (
                "AI insights are temporarily unavailable. Please ask your system "
                "administrator to configure WorkCom/OpenAI settings in Enhanced AI Settings."
            )
        }


@frappe.whitelist()
def get_status_chart() -> Dict[str, Any]:
    """Get status distribution chart data."""
    status_map, _ = get_distribution_maps()
    colors = {
        "Active": "#28a745",
        "Inactive": "#6c757d",
        "Suspended": "#ffc107",
        "Pending Verification": "#17a2b8",
        "Blacklisted": "#dc3545",
        "Unknown": "#adb5bd",
    }
    return {
        "data": {
            "labels": list(status_map.keys()),
            "datasets": [{"name": "Count", "values": list(status_map.values())}]
        },
        "type": "bar",
        "colors": [colors.get(k, "#5e64ff") for k in status_map.keys()],
    }


@frappe.whitelist()
def get_branch_chart() -> Dict[str, Any]:
    """Get branch distribution chart data (top 10)."""
    _, branch_map = get_distribution_maps()
    return {
        "data": {
            "labels": list(branch_map.keys()),
            "datasets": [{"name": "Employers", "values": list(branch_map.values())}]
        },
        "type": "bar",
        "colors": ["#7cd6fd"],
    }


@frappe.whitelist()
def get_trend_chart(months: int = 6) -> Dict[str, Any]:
    """Get trend chart data for employer registrations over the last N months."""
    labels = []
    values = []
    anchor = getdate()

    for i in range(months - 1, -1, -1):
        mend = get_first_day(add_months(anchor, -i + 1)) if i > 0 else anchor
        count = frappe.db.count("Employer", filters={
            "creation": ["<=", mend]
        })
        mstart = add_months(get_first_day(anchor), -i)
        labels.append(mstart.strftime("%b %Y"))
        values.append(int(count))

    return {
        "data": {
            "labels": labels,
            "datasets": [{"name": "Total Employers", "values": values}]
        },
        "type": "line",
        "colors": ["#7cd6fd"],
    }



